File: 01_languages/haskell/pattern_matching.hs
import Data.Char (toUpper)

data Name = Name { 
  first :: String, 
  last :: String,
  salutation :: String
} deriving (Show)

capitalize :: Name -> String
capitalize (Name fName lName _) = (upper fName) ++ " " ++ (upper lName)
  where
    upper cs = [toUpper c | c <- cs]

pythagorean :: Double -> Double -> Double
pythagorean x y = sqrt (square x + square y)
  where
    square n = n ^^ 2

distance :: (Double, Double) -> (Double, Double) -> Double
distance (0,0) (0,0) = 0
distance p1@(x1,y1) p2@(x2,y2)
  | p1 == (0,0) || p2 == (0,0) = (let (x,y) = if p1 == (0,0) then p2 else p1 in pythagorean x y)
  | otherwise = pythagorean (x2 - x1) (y2 - y1)

main :: IO ()
main = do
  let name = Name "Dhaval" "Dalal" "Mr."
  print $ name
  print $ capitalize name
  print $ distance (0,0) (0,0)
  print $ distance (3,0) (0,0)
  print $ distance (0,0) (0,4)
  print $ distance (3,0) (0,4)
  print "DONE"

File: 01_languages/erlang/pattern_matching.erl
#!/usr/bin/env escript

-record (name, {first = "", last = "", salutation = ""}).

capitalize(#name {first = FName, last = LName}) ->
	string:to_upper(FName) ++ " " ++ string:to_upper(LName).

% This version checks equality (P, P) - neat :)
distance(P, P) -> 0;
distance(P1, P2) when P1 =:= {0,0}; P2 =:= {0,0} ->
  {X,Y} = if P1 =:= {0,0} -> P2;
             P2 =:= {0,0} -> P1
  end,
  math:sqrt(math:pow(X, 2) + math:pow(Y, 2));
  
distance({X1,Y1}, {X2,Y2}) ->
  math:sqrt(math:pow(X2 - X1, 2) + math:pow(Y2 - Y1, 2)).

main([]) ->
  Name = #name {first = "Dhaval", last = "Dalal", salutation = "Mr."},
  io:format("~p~n", [Name]),
  io:format("~p~n", [capitalize(Name)]),
  io:format("~p~n", [distance({0,0}, {0,0})]),
  io:format("~p~n", [distance({3,0}, {0,0})]),
  io:format("~p~n", [distance({0,0}, {0,4})]),
  io:format("~p~n", [distance({3,0}, {0,4})]).
  